### PR TITLE
Object detection/classification with Dlib

### PR DESCRIPTION
**Pull request by cchadowitz-pf**
Supports two fixed network architectures defined in DNNStructures.h (`obj_detector`, `face_detector`) for classification/detection (returns bounding boxes and confidence, and class label if it exists).

TODO:
- [ ] Add additional parameters for dlib models

Additional flags:
- `USE_DLIB`: enables dlib backend, GPU detection is automatic. I haven't figured out how to have dlib build with CUDA but *without* cuDNN as dlib has its own FindCuda routine that automatically detects if cuDNN is available, so if `USE_DLIB` is `ON`, then `USE_CUDNN` is automatically turned on as well (and stated as such during cmake configuration)
- `USE_DLIB_CPU_ONLY`: builds dlib backend without GPU support

Example usage (can also see some examples from the tests in `ut_dlibapi.cc`):

Create face detection service (model available from [here](http://dlib.net/files/mmod_human_face_detector.dat.bz2), simply decompress it):
```bash
curl -X PUT "http://localhost:8080/services/face" -d '
{
  "mllib":"dlib",
  "description":"face",
  "type":"supervised",
  "parameters":{
    "input":{
      "connector":"image",
      "width": 1024,
      "height": 1024
    },
    "mllib":{
      "model_type": "face_detector"
    }
  },
  "model":{
    "repository":"/opt/models/face/"
  }
}'
```

Classify/detect/predict with the face detection service:
```bash
curl -X POST "http://localhost:8080/predict" -d '{
    "service":"face",
    "parameters":{
      "output":{
        "bbox": true
      }
    },
    "data":["https://upload.wikimedia.org/wikipedia/commons/6/68/Dwayne_Johnson_at_the_2009_Tribeca_Film_Festival.jpg", "https://www.deepdetect.com/models/tf/grace_hopper.jpg"]
  }'
```
which should return something like:
```json
{
  "status": {
    "code": 200,
    "msg": "OK"
  },
  "head": {
    "method": "/predict",
    "service": "face",
    "time": 3189
  },
  "body": {
    "predictions": [
      {
        "classes": [
          {
            "prob": 1.0872774124145508,
            "last": true,
            "bbox": {
              "xmax": 700,
              "ymax": 215,
              "ymin": 623,
              "xmin": 293
            },
            "cat": ""
          }
        ],
        "uri": "https://upload.wikimedia.org/wikipedia/commons/6/68/Dwayne_Johnson_at_the_2009_Tribeca_Film_Festival.jpg"
      },
      {
        "classes": [
          {
            "prob": 1.0895934104919434,
            "last": true,
            "bbox": {
              "xmax": 658,
              "ymax": 249,
              "ymin": 532,
              "xmin": 375
            },
            "cat": ""
          }
        ],
        "uri": "https://www.deepdetect.com/models/tf/grace_hopper.jpg"
      }
    ]
  }
}
```

Create object (vehicle front/rear) detection service (model available from [here](http://dlib.net/files/mmod_front_and_rear_end_vehicle_detector.dat.bz2), simply decompress it):
```bash
curl -X PUT "http://localhost:8080/services/vehicle" -d '
{
  "mllib":"dlib",
  "description":"vehicle",
  "type":"supervised",
  "parameters":{
    "input":{
      "connector":"image",
      "width": 0,
      "height": 0
    },
    "mllib":{
      "model_type": "obj_detector"
    }
  },
  "model":{
    "repository":"/opt/models/vehicle/"
  }
}'
```
Using `"width":0, "height": 0` maintains the native resolution (**Note:** this will cause errors if batching images of different resolutions)

Classify/detect/predict with the vehicle front/rear detection service:
```bash
curl -X POST "http://localhost:8080/predict" -d '{
    "service":"vehicle",
    "parameters":{
      "output":{
        "bbox": true
      }
    },
    "data":["https://raw.githubusercontent.com/davisking/dlib/master/examples/mmod_cars_test_image2.jpg"]
  }'
```
which should return something like:
```json
{
  "status": {
    "code": 200,
    "msg": "OK"
  },
  "head": {
    "method": "/predict",
    "service": "vehicle",
    "time": 618
  },
  "body": {
    "predictions": [
      {
        "classes": [
          {
            "prob": 1.2941689491271973,
            "bbox": {
              "xmax": 1871,
              "ymax": 237,
              "ymin": 343,
              "xmin": 1728
            },
            "cat": "rear"
          },
          {
            "prob": 1.0142754316329956,
            "bbox": {
              "xmax": 1371,
              "ymax": 239,
              "ymin": 421,
              "xmin": 1124
            },
            "cat": "rear"
          },
          {
            "prob": 0.9298916459083557,
            "bbox": {
              "xmax": 2087,
              "ymax": 210,
              "ymin": 393,
              "xmin": 1840
            },
            "cat": "rear"
          },
          {
            "prob": 0.8713066577911377,
            "bbox": {
              "xmax": 1739,
              "ymax": 218,
              "ymin": 306,
              "xmin": 1619
            },
            "cat": "rear"
          },
          {
            "prob": 0.849409818649292,
            "bbox": {
              "xmax": 1486,
              "ymax": 202,
              "ymin": 253,
              "xmin": 1417
            },
            "cat": "rear"
          },
          {
            "prob": 0.578871488571167,
            "bbox": {
              "xmax": 1086,
              "ymax": 218,
              "ymin": 269,
              "xmin": 1017
            },
            "cat": "front"
          },
          {
            "prob": 0.4774478077888489,
            "bbox": {
              "xmax": 363,
              "ymax": 211,
              "ymin": 363,
              "xmin": 157
            },
            "cat": "front"
          },
          {
            "prob": 0.2267676591873169,
            "bbox": {
              "xmax": 740,
              "ymax": 210,
              "ymin": 393,
              "xmin": 493
            },
            "cat": "front"
          },
          {
            "prob": 0.19519296288490295,
            "bbox": {
              "xmax": 1366,
              "ymax": 194,
              "ymin": 245,
              "xmin": 1297
            },
            "cat": "rear"
          },
          {
            "prob": 0.1648062765598297,
            "bbox": {
              "xmax": 1182,
              "ymax": 202,
              "ymin": 253,
              "xmin": 1113
            },
            "cat": "front"
          },
          {
            "prob": 0.02589753270149231,
            "bbox": {
              "xmax": 1610,
              "ymax": 213,
              "ymin": 275,
              "xmin": 1527
            },
            "cat": "rear"
          },
          {
            "prob": 0.01753876358270645,
            "last": true,
            "bbox": {
              "xmax": 1342,
              "ymax": 202,
              "ymin": 253,
              "xmin": 1273
            },
            "cat": "rear"
          }
        ],
        "uri": "https://raw.githubusercontent.com/davisking/dlib/master/examples/mmod_cars_test_image2.jpg"
      }
    ]
  }
}
```

### CODE REVIEW

File: src/backends/dlib/dlibinputconns.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        void transform(const APIData &ad) {
+            try {
+                ImgInputFileConn::transform(ad);
+            }
+            catch (InputConnectorBadParamException &e) {
+                throw;
+            }
+            for (size_t i = 0; i < _images.size(); i++) {
+                dlib::matrix<dlib::rgb_pixel> inputImg;
+                dlib::assign_image(inputImg, dlib::cv_image<dlib::rgb_pixel>(_images.at(i)));
```
Comment (chichaj) [inline src/backends/dlib/dlibinputconns.h]: Do you believe you should add _bw support ?

File: src/backends/dlib/dliblib.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+        int predict(const APIData &ad, APIData &out);
+
+
+    public:
+        // general parameters
+
+        std::string _net_type; // model type
+        // model, depending on type specified
+        net_type_objDetector objDetector;
```
Comment (chichaj) [inline src/backends/dlib/dliblib.h]: Let's add _ to the class variables ?

File: src/imginputfileconn.h [hunk 3]
```diff
@@ -78,9 +78,23 @@ namespace dd
                                      _unchanged_data ? CV_LOAD_IMAGE_UNCHANGED :
                                      (_bw ? CV_LOAD_IMAGE_GRAYSCALE : CV_LOAD_IMAGE_COLOR)));
 	_imgs_size.push_back(std::pair<int,int>(img.rows,img.cols));
-	cv::Size size(_width,_height);
-	cv::Mat rimg;
-	cv::resize(img,rimg,size,0,0,CV_INTER_CUBIC);
+    cv::Mat rimg;
+	if (_width == 0 || _height == 0) {
```
Comment (chichaj) [inline src/imginputfileconn.h]: Ok, it doesn't seem to affect our other backends

File: src/backends/tf/tfmodel.cc [hunk 4]
```diff
@@ -44,7 +44,7 @@ namespace dd
   	int e = fileops::list_directory(repo,true,false,lfiles);
     if (e != 0)
       {
-	logger->error("error reading or listing caffe models in repository {}",repo);
+	logger->error("error reading or listing tensorflow models in repository {}",repo);
```
Comment (chichaj) [inline src/backends/tf/tfmodel.cc]: Thanks !

File: src/backends/dlib/dlibinputconns.h (hunk 1 above)
Comment (cchadowitz-pf) [reply src/backends/dlib/dlibinputconns.h]: Good idea, thanks. I'll add that in.

Comment (cchadowitz-pf) [reply src/backends/dlib/dlibinputconns.h]: In the meanwhile I added support for `_bw` in that if it is specified, it is correctly converted from the `cv::Mat` to a `dlib::matrix` object such that the supported dlib models can accept the image.

Comment (beniz) [reply src/backends/dlib/dlibinputconns.h]: Yes I believe this is the right behavior for now. Usually the issue is that the image channels need to match the network's input size. Thanks for the code update.

Comment (cchadowitz-pf) [reply src/backends/dlib/dlibinputconns.h]: Got it 👍 minor tweak to remove unnecessary code but should be good now.

**Comment by beniz** [comment]
Thanks for this contribution!

**Pull request merged**